include/evmc/mocked_host.hpp

```diff
@@ -0,0 +1,320 @@
+                                    const bytes32& value) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return EVMC_STORAGE_UNCHANGED;
+
+        auto& old = it->second.storage[key];
+
+        // Follow https://eips.ethereum.org/EIPS/eip-1283 specification.
+        // WARNING! This is not complete implementation as refund is not handled here.
+
+        if (old.value == value)
+            return EVMC_STORAGE_UNCHANGED;
+
+        evmc_storage_status status;
+        {
+            if (!old.dirty)
+            {
+                old.dirty = true;
+                if (!old.value)
+                    status = EVMC_STORAGE_ADDED;
+                else if (value)
+                    status = EVMC_STORAGE_MODIFIED;
+                else
+                    status = EVMC_STORAGE_DELETED;
+            }
+            else
+                status = EVMC_STORAGE_MODIFIED_AGAIN;
+        }
+
+        old.value = value;
+        return status;
+    }
+
+    /// Get the account's balance (EVMC Host method).
+    uint256be get_balance(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return {};
+
+        return it->second.balance;
+    }
+
+    /// Get the account's code size (EVMC host method).
+    size_t get_code_size(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return 0;
+        return it->second.code.size();
+    }
+
+    /// Get the account's code hash (EVMC host method).
+    bytes32 get_code_hash(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return {};
+        return it->second.codehash;
+    }
+
+    /// Copy the account's code to the given buffer (EVMC host method).
+    size_t copy_code(const address& addr,
+                     size_t code_offset,
+                     uint8_t* buffer_data,
+                     size_t buffer_size) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return 0;
+
+        const auto& code = it->second.code;
+
+        if (code_offset >= code.size())
+            return 0;
+
+        const auto n = std::min(buffer_size, code.size() - code_offset);
+
+        if (n > 0)
+            std::copy_n(&code[code_offset], n, buffer_data);
+        return n;
+    }
+
+    /// Selfdestruct the account (EVMC host method).
+    void selfdestruct(const address& addr, const address& beneficiary) noexcept override
+    {
+        record_account_access(addr);
+        recorded_selfdestructs.push_back({addr, beneficiary});
```

Comment:
could be `emplace_back`

Reply:
`emplace_back` does not work with `{}` — I believe it needs to know the type.